### PR TITLE
Make V.graph properly typed

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114160
* #114162
* #114161
* __->__ #114025
* #114013

Previously it lacked a type hint and so was treated as an Any type. This
resulted in a lot of untyped code downstream as V.graph is referenced in
many places in inductor code. I've typed it properly now as
GraphLowering, and fixed the numerous type errors this surfaced.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -193,37 +197,34 @@ def __init__(
...
         self.never_reuse_buffers: Set[str] = set()
         self.inplaced_to_remove: Set[str] = set()
-        self.wrapper_code: Optional[WrapperCodeGen] = None
         # See `ProxyExecutor Design Note` in ir.py for more details
         self.extern_kernel_nodes: List[ir.ExternKernelNode] = []
         self.extern_node_serializer: Optional[
             Callable[[List[ir.ExternKernelNode]], Any]
         ] = extern_node_serializer
-        self.current_node: Optional[torch.fx.Node] = None
+        self.current_node: torch.fx.Node = None  # type: ignore[assignment]
```
Comment (int3) [inline torch/_inductor/graph.py:216]: type-ignore added as the `Optional` type resulted in too many `assert` statements

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1932,7 +1939,7 @@ def codegen_kernel(self, name=None):
             if tree.prefix != "r" or self.inside_reduction:
                 sizearg = SizeArg(f"{tree.prefix}numel", tree.numel)
                 signature.append(sizearg)
-                triton_meta["signature"][len(argdefs)] = signature_of(
+                triton_meta_signature[len(argdefs)] = signature_of(
```
Comment (int3) [inline torch/_inductor/codegen/triton.py:1942]: `mypy` sees `triton_meta["signature"]` as being of `object` type as it unifies the types of all of `triton_meta`'s values. Thus we use a local variable to preserve the type of the signature value

**Review by eellison — APPROVED** [review]
nice!!

**Pull request closed without merging**